src/app/api/instagram/syncBookmarks/route.ts

```diff
@@ -0,0 +1,261 @@
+import { apiError, apiWarn } from "@/lib/api-helpers/response";
+import { MAIN_TABLE_NAME } from "@/utils/constants";
+
+const ROUTE = "instagram-sync-bookmarks";
+
+const InstagramSyncBookmarksPayloadSchema = z.object({
+	data: z.array(
+		z.object({
+			description: z.string(),
+			ogImage: z.string().nullable(),
+			title: z.string(),
+			type: z.string(),
+			url: z.string(),
+			meta_data: z.object({
+				instagram_username: z.string(),
+				instagram_profile_pic: z.string().nullable().optional(),
+				favIcon: z.string(),
+				video_url: z.string().nullable().optional(),
+				saved_collection_names: z.array(z.string()).optional(),
+			}),
+			inserted_at: z.string().datetime().optional(),
+			sort_index: z.string(),
+			category_name: z.string().optional(),
+		}),
+	),
+});
+
+const InstagramSyncBookmarksResponseSchema = z.object({
+	success: z.boolean(),
+	error: z.string().nullable(),
+	data: z
+		.array(
+			z.object({
+				category_id: z.number(),
+				description: z.string().nullable(),
+				id: z.number(),
+				inserted_at: z.string(),
+				meta_data: z.unknown().nullable(),
+				ogImage: z.string().nullable(),
+				screenshot: z.string().nullable(),
+				sort_index: z.string().nullable(),
+				title: z.string().nullable(),
+				trash: z.boolean(),
+				type: z.string().nullable(),
+				url: z.string().nullable(),
+				user_id: z.string(),
+			}),
+		)
+		.optional(),
+});
+
+export type InstagramSyncBookmarksPayload = z.infer<
+	typeof InstagramSyncBookmarksPayloadSchema
+>;
+
+export type InstagramSyncBookmarksResponse = z.infer<
+	typeof InstagramSyncBookmarksResponseSchema
+>;
+
+export const POST = createSupabasePostApiHandler({
+	route: ROUTE,
+	inputSchema: InstagramSyncBookmarksPayloadSchema,
+	outputSchema: InstagramSyncBookmarksResponseSchema,
+	handler: async ({ data, supabase, user, route }) => {
+		const userId = user.id;
+
+		console.log(`[${route}] API called:`, {
+			userId,
+			bookmarkCount: data.data.length,
+		});
+
+		// adding user_id in the data to be inserted
+		const insertData = data.data.map((item) => ({
+			...item,
+			user_id: userId,
+		}));
+
+		// get the urls who are instagram posts present in table, we fetch only the urls there are there in the insertData for the query optimization
+		const { data: duplicateCheckData, error: duplicateCheckError } =
+			await supabase
+				.from(MAIN_TABLE_NAME)
+				.select("url")
+				// get only the urls that are there in the payload
+				.in(
+					"url",
+					insertData.map((item) => item.url),
+				)
+				.eq("user_id", userId)
+				.eq("type", "instagram");
+
+		if (duplicateCheckError) {
+			console.warn(`[${route}] DB duplicateCheckError`, duplicateCheckError);
+
+			return apiError({
+				route,
+				message: `DB duplicateCheckError: ${duplicateCheckError.message}`,
+				error: duplicateCheckError,
+				operation: "duplicate_check",
+				userId,
+			});
+		}
+
+		// filter out the duplicates from the payload data
```

Comment:
🔧 **O(n²) Complexity** - Convert to Set for O(n) lookup. See [detailed analysis](https://github.com/timelessco/recollect/pull/718#issuecomment-3698914597)